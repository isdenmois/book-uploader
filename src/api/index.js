import cheerio from 'react-native-cheerio';
import { FLIBUSTA_HOST, ZLIB_HOST, USER_AGENT } from '@env';

const flibusta = {
  host: FLIBUSTA_HOST,
  useProxy: true,
  search: {
    path: '/api/rewrite',
    params: { path: '/opds/search', searchType: 'books' },
    query: {
      type: 'query',
      name: 'searchTerm',
    },
  },
  selectors: {
    entry: 'entry',
    link: 'link[href$="fb2"]',
    title: 'title',
    author: 'author name',
    translation: {
      selector: 'content',
      regexp: 'Перевод:\\s?(.+?)\\s?[&<]',
    },
    lang: {
      selector: 'content',
      regexp: 'Язык:\\s?(.+?)\\s?[&<]',
    },
    size: {
      selector: 'content',
      regexp: 'Размер:\\s?(.+?)\\s?[&<]',
    },
    ext: 'fb2.zip',
  },
};

const zlib = {
  host: ZLIB_HOST,
  search: {
    path: '/s/{{name}}/',
    params: { e: 1, language: '', extension: 'epub' },
    query: {},
  },
  selectors: {
    entry: '#searchResultBox .resItemBox',
    link: 'h3[itemprop="name"] a',
    title: 'h3[itemprop="name"] a',
    author: '.authors a',
    lang: {
      selector: '.property_language',
      replace: 'Language:',
    },
    size: {
      selector: '.property__file .property_value',
      replace: 'EPUB, ',
    },
    ext: 'epub',
  },
};

function flibustaFileUrl(link) {
  return `${FLIBUSTA_HOST}/api/rewrite?path=${encodeURIComponent(link)}`;
}

async function zlibFileUrl(link) {
  const body = await fetch(`${ZLIB_HOST}${link}`, { headers }).then(r => r.text());
  const $ = cheerio.load(body);

  return ZLIB_HOST + $(`a.addDownloadedBook`).attr('href');
}

export function fileUrl(type, link) {
  if (type === 'zlib') {
    return zlibFileUrl(link);
  }

  return flibustaFileUrl(link);
}

const providers = {
  flibusta,
  zlib,
};

function querystring(query = {}) {
  // get array of key value pairs ([[k1, v1], [k2, v2]])
  const qs = Object.entries(query)
    // filter pairs with undefined value
    .filter(pair => pair[1] !== undefined)
    // encode keys and values, remove the value if it is null, but leave the key
    .map(pair =>
      pair
        .filter(i => i !== null)
        .map(encodeURIComponent)
        .join('='),
    )
    .join('&');

  return qs && '?' + qs;
}

function parseSearch(body, selectors) {
  const $ = cheerio.load(body);
  const translation = selectors.translation && new RegExp(selectors.translation.regexp, 'i');

  const books = $(selectors.entry).map((key, entry) => {
    entry = $(entry);
    const data = {
      title: entry.find(selectors.title).text(),
      link: entry.find(selectors.link).attr('href'),
      ext: selectors.ext,
    };

    const authors = $(entry).find(selectors.author);

    if (authors.length > 1) {
      data.authors = authors
        .map((i, a) => $(a).text())
        .get()
        .join('; ');
    } else {
      data.authors = authors.text();
    }

    if (translation) {
      data.translation = entry.find(selectors.translation.selector).text()?.match(translation)?.[1];
    }
    const lang = entry.find(selectors.lang.selector);

    if (selectors.lang.regexp) {
      data.lang = lang.text()?.match(selectors.lang.regexp)?.[1];
    } else {
      data.lang = lang.text().replace(selectors.lang.replace, '').trim();
    }

    const size = entry.find(selectors.size.selector);

    if (selectors.size.regexp) {
      data.size = size.text()?.match(selectors.size.regexp)?.[1];
    } else {
      data.size = size.text().replace(selectors.size.replace, '').trim();
    }

    return data;
  });

  return books.get();
}

const headers = {
  'User-Agent': USER_AGENT,
};

export async function searchHandler(type, name) {
  const config = providers[type] || flibusta;

  const search = config.search;

  let url = config.host + search.path;

  const params = { ...search.params };

  if (search.query.type === 'query') {
    params[search.query.name] = name;
  } else {
    url = url.replace('{{name}}', name);
  }

  url = url + querystring(params);

  try {
    const response = await fetch(url, { headers, params, proxy: process.env.HTTPS_PROXY });
    const body = await response.text();

    return { data: parseSearch(body, config.selectors, config.fileUrl) };
  } catch (e) {
    return { error: (e && e.message) || e };
  }
}
