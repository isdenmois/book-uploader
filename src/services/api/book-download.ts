import cheerio from 'react-native-cheerio';
import RNFS from 'react-native-fs';
import { TOR_HOST, FLIBUSTA_HOST, ZLIB_HOST, USER_AGENT } from '@env';
import AsyncStorage from '@react-native-community/async-storage';
import { querystring } from './utils';
import * as tor from './tor-request';
import { ZLIB_COOKIE } from './login';
import { BookItem, ProviderType } from './types';

type SetProgressFn = (progress: number) => void;
export async function downloadFile(file: BookItem, fileName: string, setProgress: SetProgressFn): Promise<void> {
  setProgress(0.01);
  const { host, path, query, headers } = await getUrl(file.type, file.link);

  // TODO: move to tor-request
  await RNFS.downloadFile({
    fromUrl: TOR_HOST + '/api/rewrite' + querystring(Object.assign({}, query, { host, path })),
    toFile: `${RNFS.DocumentDirectoryPath}/${fileName}`,
    headers: Object.assign({}, headers, { 'user-agent': USER_AGENT }),
    progress: ({ contentLength, bytesWritten }) => setProgress(Math.round((bytesWritten / contentLength) * 100)),
  }).promise;

  setProgress(0);
}

export function getUrl(type: ProviderType, link: string): any {
  if (type === 'zlib') {
    return zlibFileUrl(link);
  }

  return flibustaFileUrl(link);
}

function flibustaFileUrl(link: string) {
  return { host: FLIBUSTA_HOST, path: link };
}

async function zlibFileUrl(link: string) {
  const headers = { Cookie: await AsyncStorage.getItem(ZLIB_COOKIE) };
  const $ = cheerio.load(await tor.request(ZLIB_HOST, link, { headers }));
  const path = $('a.addDownloadedBook').attr('href');
  const query = { nofollow: true };

  return { host: ZLIB_HOST, path, query, headers };
}
