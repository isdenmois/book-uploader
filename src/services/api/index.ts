export { downloadFile } from './book-download';
export { bookSearch } from './book-search';
export { sendLogin, setCookie } from './login';
export { uploadBook } from './upload';
