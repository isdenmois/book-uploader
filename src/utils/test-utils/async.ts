import { renderHook, RenderHookOptions, RenderHookResult } from '@testing-library/react-hooks';
import { act } from '@testing-library/react-native';
import { create, ReactTestRenderer } from 'react-test-renderer';

/*global jest */

type FunctionPropertyNames<T> = { [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never }[keyof T] & string;

export const mockPromise = <T extends {}, M extends FunctionPropertyNames<Required<T>>>(object: T, method: M) => {
  let resolvePromise;
  const promise: any = new Promise(resolve => (resolvePromise = resolve));

  if (method in object) {
    jest.spyOn(object, method).mockReturnValue(promise);
  } else {
    (object[method] as any) = () => promise;
  }

  const resolve = value => act(async () => resolvePromise(value));

  return [resolve] as const;
};

export const mockPromiseValue = (obj, prop, value) => {
  const promise = Promise.resolve(value);

  if (prop in obj) {
    jest.spyOn(obj, prop).mockReturnValue(promise);
  } else {
    obj[prop] = () => promise;
  }
};

export const createAsync = async (elem): Promise<ReactTestRenderer> => {
  let result: ReactTestRenderer;

  await act(async () => (result = create(elem)));

  return result;
};

export const immediate = () => new Promise(resolve => setImmediate(resolve));

export async function renderHookAsync<P, R>(
  callback: (props: P) => R,
  options?: RenderHookOptions<P>,
): Promise<RenderHookResult<P, R>> {
  let result: RenderHookResult<P, R>;

  await act(async () => (result = renderHook(callback, options)));

  return result;
}

export function mock<T extends {}, M extends FunctionPropertyNames<Required<T>>>(object: T, method: M) {
  if (method in object) {
    return jest.spyOn(object, method);
  }

  return ((object[method] as any) = () => jest.fn());
}
