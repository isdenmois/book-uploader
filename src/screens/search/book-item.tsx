import React, { useState } from 'react';
import { slugify } from 'transliteration';
import { Alert, Text, View, TouchableOpacity, ToastAndroid, StyleSheet, ViewStyle } from 'react-native';
import * as api from 'services/api/book-download';

export function BookItem({ item }) {
  const [progress, onPress] = useDownload(item);

  return (
    <TouchableOpacity style={s.container} onPress={onPress} testID='book-item'>
      <Text>{item.title}</Text>
      <Text>{item.authors}</Text>
      {!!item.lang && <Text>Lang: {item.lang}</Text>}
      {!!item.translation && <Text>Translator: {item.translation}</Text>}
      {!!item.size && <Text>Size: {item.size}</Text>}
      {progress > 0 && (
        <View
          style={{
            height: 5,
            position: 'relative',
            backgroundColor: '#ddd',
            borderRadius: 2,
            marginTop: 4,
            overflow: 'hidden',
          }}
        >
          <View
            style={{ width: `${progress}%`, backgroundColor: '#f00', position: 'absolute', left: 0, top: 0, bottom: 0 }}
          />
        </View>
      )}
    </TouchableOpacity>
  );
}

export function useDownload(file) {
  const [progress, setProgress] = useState(0);
  const title = slugify(file.title).slice(0, 100);
  const fileName = `${title}.${file.ext}`;

  const onPress = useConfirm(fileName, 'Start download?', async () => {
    ToastAndroid.show('Start downloading', ToastAndroid.SHORT);

    try {
      await api.downloadFile(file, fileName, setProgress);

      ToastAndroid.show(`File ${fileName} has downloaded!`, ToastAndroid.SHORT);
    } catch (e) {
      console.error(e?.message || e);
      ToastAndroid.show(`Error: ${e?.message || e}`, ToastAndroid.SHORT);
    }

    setProgress(0);
  });

  return [progress, onPress] as const;
}

export function useConfirm(title, message, onSuccess) {
  return () => {
    confirm(title, message, onSuccess);
  };
}

function confirm(title: string, message: string, onSuccess) {
  Alert.alert(title, message, [
    {
      text: 'Cancel',
      style: 'cancel',
    },
    { text: 'OK', onPress: onSuccess, style: 'destructive' },
  ]);
}

const s = StyleSheet.create({
  container: {
    marginBottom: 10,
  } as ViewStyle,
});
