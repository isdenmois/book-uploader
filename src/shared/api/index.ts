import { downloadFile } from './book-download'
import { getDailyDownloads, DaylyDownloadsStat } from './daily-downloads'
import { bookSearch } from './book-search'
import { sendLogin, setCookie } from './login'
import { uploadFile } from './upload'

export * from './types'
export type { DaylyDownloadsStat }
export const api = { downloadFile, bookSearch, getDailyDownloads, sendLogin, setCookie, uploadFile }
