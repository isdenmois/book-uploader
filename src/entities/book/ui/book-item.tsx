import React, { FC, memo } from 'react'
import { StyleSheet } from 'react-native'
import FastImage from 'react-native-fast-image'

import { BookItem as IBookItem } from 'shared/api'
import { getImageUrl } from 'shared/api/book-image'
import { Box, FileIcon, Item, Text } from 'shared/ui'

interface Props {
  item: IBookItem
  onPress: (item: IBookItem) => void
}

export const BookItem: FC<Props> = memo<Props>(
  ({ item, onPress }) => {
    const other = [item.size, item.lang, item.translation].filter(p => p).join(', ')
    const ext = item.ext.replace('.zip', '')
    const imageUrl = getImageUrl(item.type, item.imageUrl)

    return (
      <Item mb={3} onPress={() => onPress(item)} testID='book-item'>
        {!!imageUrl && (
          <Box borderRadius={4} width={40} height={50} overflow='hidden' backgroundColor='searchSelected'>
            <FastImage style={s.image} source={{ uri: imageUrl }} resizeMode='cover' />

            <Text fontSize={12} fontWeight='bold' textAlign='center' color='white' style={s.text}>
              {ext}
            </Text>
          </Box>
        )}
        {!imageUrl && <FileIcon color='searchSelected' text={ext} />}

        <Item.Text suptitle={item.authors} title={item.title} subtitle={other} />
      </Item>
    )
  },
  (prev, next) => prev.item === next.item,
)

const s = StyleSheet.create({
  image: {
    width: 40,
    height: 50,
  },
  text: {
    backgroundColor: '#0005',
    bottom: 0,
    left: 0,
    right: 0,
    position: 'absolute',
  },
})
