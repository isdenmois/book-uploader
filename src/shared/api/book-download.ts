import cheerio from 'react-native-cheerio'
import RNFS from 'react-native-fs'
import { API_CONFIG } from './config'
import { MMKV } from 'shared/libs'
import { querystring } from './utils'
import * as tor from './tor-request'
import { ZLIB_COOKIE } from './login'
import { BookItem, ProviderType } from './types'

type SetProgressFn = (progress: number) => void
export async function downloadFile(file: BookItem, fileName: string, setProgress: SetProgressFn): Promise<void> {
  setProgress(0.01)
  const { host, path, query, headers } = await getUrl(file.type, file.link)

  // TODO: move to tor-request
  await RNFS.downloadFile({
    fromUrl: API_CONFIG.TOR_HOST + '/api/rewrite' + querystring(Object.assign({}, query, { host, path })),
    toFile: `${RNFS.DocumentDirectoryPath}/${fileName}`,
    headers: Object.assign({}, headers, { 'user-agent': API_CONFIG.USER_AGENT }),
    progress: ({ contentLength, bytesWritten }) => setProgress(Math.round((bytesWritten / contentLength) * 100)),
  }).promise

  setProgress(0)
}

export function getUrl(type: ProviderType, link: string): any {
  if (type === 'ZLIB') {
    return zlibFileUrl(link)
  }

  return flibustaFileUrl(link)
}

function flibustaFileUrl(link: string) {
  return { host: API_CONFIG.FLIBUSTA_HOST, path: link }
}

async function zlibFileUrl(link: string) {
  const headers = { Cookie: MMKV.getString(ZLIB_COOKIE) }
  const $ = cheerio.load(await tor.request(API_CONFIG.ZLIB_HOST, link, { headers }))
  const path = $('a.addDownloadedBook').attr('href')
  const query = { nofollow: true }

  return { host: API_CONFIG.ZLIB_HOST, path, query, headers }
}
