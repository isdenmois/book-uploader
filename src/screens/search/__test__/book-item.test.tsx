import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Alert, ToastAndroid } from 'react-native';
import { mockPromise } from 'utils/test-utils/async';
jest.mock('services/api/book-download', () => ({}));
import * as api from 'services/api/book-download';
import { BookItem } from '../book-item';
import { act } from 'react-test-renderer';

test('BookItem', async () => {
  const [resolve, , downloadFile] = mockPromise(api, 'downloadFile');
  const alert = jest.spyOn(Alert, 'alert');
  const show = jest.spyOn(ToastAndroid, 'show').mockResolvedValue(undefined as never);
  const file = {
    title: 'super title',
    author: 'uauthr',
    lang: 'ru',
    translation: 'someone',
    size: '100500',
    ext: 'epub',
  };

  const { toJSON, getByTestId } = render(<BookItem item={file} />);

  expect(toJSON()).toMatchSnapshot();
  expect(downloadFile).not.toHaveBeenCalled();

  fireEvent.press(getByTestId('book-item'));
  expect(alert).toHaveBeenCalledTimes(1);
  expect(show).not.toHaveBeenCalled();

  alert.mock.calls[0][2][1].onPress(); // onSuccess;
  expect(show).toHaveBeenCalledWith('Start downloading', ToastAndroid.SHORT);
  expect(downloadFile).toHaveBeenCalledTimes(1);

  const setProgress = downloadFile.mock.calls[0][2];
  act(() => setProgress(50));

  expect(toJSON()).toMatchSnapshot();

  await resolve(null);

  expect(show).toHaveBeenCalledWith('File super-title.epub has downloaded!', ToastAndroid.SHORT);
});
