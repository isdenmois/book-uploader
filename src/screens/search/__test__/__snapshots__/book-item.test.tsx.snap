// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookItem 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "marginBottom": 10,
      "opacity": 1,
    }
  }
  testID="book-item"
>
  <Text>
    super title
  </Text>
  <Text />
  <Text>
    Lang: 
    ru
  </Text>
  <Text>
    Translator: 
    someone
  </Text>
  <Text>
    Size: 
    100500
  </Text>
</View>
`;

exports[`BookItem 2`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "marginBottom": 10,
      "opacity": 1,
    }
  }
  testID="book-item"
>
  <Text>
    super title
  </Text>
  <Text />
  <Text>
    Lang: 
    ru
  </Text>
  <Text>
    Translator: 
    someone
  </Text>
  <Text>
    Size: 
    100500
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#ddd",
        "borderRadius": 2,
        "height": 5,
        "marginTop": 4,
        "overflow": "hidden",
        "position": "relative",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#f00",
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "50%",
        }
      }
    />
  </View>
</View>
`;
